### PR TITLE
E2E composability testing

### PR DESCRIPTION
**Pull request by mori360**
Add 3D(pp+tp+fsdp) test `test_3d_with_tp_dp_pp` at test_pp_compodability
Currently provide @parametrize on 
"ScheduleClass" for pp in [ScheduleGPipe, Schedule1F1B, ScheduleInterleaved1F1B, ScheduleLoopedBFS, ScheduleInterleavedZeroBubble]
"MixedPrecisionParam" for fsdp in [torch.bfloat16, torch.float32]

Future work:
1. add fp8
2. add cp(context parallelism) to enable 4D test

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

### CODE REVIEW

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+                [param for model in models for param in model.parameters()],
+                1.0,
+                foreach=True,
+                pp_mesh=pp_mesh if parallel_dims.pp_enabled else None,
+            )
+
+            for optimizer in optimizers:
+                optimizer.step()
+
+    def pipeline_llama_manual_split(
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_3d_composability.py:208]: suggest you delete the 2 helpers 'pipeline_llama_manual_split' and 'generate_split_points'.

1. it is complicated code to read, and it is complicated _because_ it is supposed to work on a real llama model and be flexible to cmdline for different split points
2. we do not need flexibility in the unit test and we are not using a llama model.  There is no `.tok_embeddings` field in the model
3. it would be useful to test the helper functions _in_ torchtitan, but we already do that.  copying them and then testing the copy is not useful in and of itself.

Conclusion- better to just construct N MLPModel()  objects and assign them to PipelineStage classes yourself.  It should be easy to do in 5 lines of code instead of 90

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+                # Middle stages get an extra layer if there are any remaining
+                if extra_layers > 0:
+                    current_layer += base_interval + 1
+                    extra_layers -= 1
+                else:
+                    current_layer += base_interval
+            splits.append("layers." + str(current_layer))
+        return splits
+
+    def apply_tp(
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_3d_composability.py:298]: it feels like apply_tp is not really doing anything. It will not find any 'tok_embeddings' or 'norm' or 'output' in the MLPModule. 

For both apply_tp and apply_fsdp, the important part is NOT to use the same helpers as torchtitan, but to do something that works correctly.

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+                    total_norm, op=dist.ReduceOp.SUM, group=pp_mesh.get_group()
+                )
+                total_norm **= 1.0 / norm_type
+
+        torch.nn.utils.clip_grads_with_norm_(parameters, max_norm, total_norm, foreach)
+        return total_norm
+
+
+@dataclass
+class ParallelDims:
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_3d_composability.py:457]: I suspect we do not need to copy this class from TorchTitan.  It is a nice helper but in the unit test case you can probably just construct your mesh directly.

**Review by wconstab — CHANGES_REQUESTED** [review]
Thanks for working on this.  There are a few issues that need to be addressed, with comments inline.  The main themes are 
- parallelisms aren't applied correctly, since the test uses MLPModel but the helpers expect llama model
- the organization can be improved- combine with other composability tests, reduce copied helpers

File: .ci/pytorch/multigpu-test.sh [hunk 4]
```diff
@@ -52,6 +52,7 @@ time python test/run_test.py --verbose -i distributed/_composable/fsdp/test_full
 
 # ND composability tests
 time python test/run_test.py --verbose -i distributed/_composable/test_composability/test_2d_composability
+time python test/run_test.py --verbose -i distributed/_composable/test_composability/test_3d_composability
```
[the remark this answers is not in the document]
Comment (wconstab) [reply .ci/pytorch/multigpu-test.sh:55]: to keep things simple, i suggest (1) in this PR just add new test to test_pp_composability.py. (2) in another PR, move test_pp_composability to `test/distributed/test_composability.py`.

I'm less sure what to do with existing tests inside `test/distributed/_composable/` folder, but we can discuss that later.

File: test/distributed/_composable/test_composability/test_3d_composability.py [hunk 5]
```diff
@@ -0,0 +1,565 @@
...
+
+
+class Test3DTraining(FSDPTest):
+    global num_layers
+    num_layers = 8
+
+    @skip_if_lt_x_gpu(8)
+    @skipIfRocm
+    def test_3d(self):
+        self.run_subtests(
```
[the remark this answers is not in the document]
Comment (mori360) [reply test/distributed/_composable/test_composability/test_3d_composability.py:44]: would switch to use `@parameterize` in the next version

File: .ci/pytorch/multigpu-test.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (mori360) [reply .ci/pytorch/multigpu-test.sh:55]: We can discuss moving the tests in the next PR

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 6]
```diff
@@ -357,6 +377,172 @@ def _dcp_test(self):
...
+        full_model = nn.ModuleList([MLPModuleEven() for _ in range(total_layers)])
+        ref_model = nn.Sequential(*copy.deepcopy(full_model))
+        ref_model.to(self.device)
+
+        # dummy loss needed just to force backwards to run in schedule step
+        def loss_fn(y, target):
+            return y.sum()
+
+        # Apply DP to stage module
+        def apply_dp(partial_model):
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_pp_composability.py:429]: nit: you could delete both comments about apply dp, and just name the fn `apply_fsdp`

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 7]
```diff
@@ -357,6 +377,172 @@ def _dcp_test(self):
...
+            # apply FSDP
+            mp_policy = MixedPrecisionPolicy(
+                param_dtype=torch.float32,
+                reduce_dtype=torch.float32,
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_pp_composability.py:433]: wondering if we should include bf16 mixed precision as one of the test configs via another parametrization.  I think it can interact with PP in some way but I forgot the details.  maybe only relevant for tracer-pp not manual-pp? need to double check

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 8]
```diff
@@ -357,6 +377,172 @@ def _dcp_test(self):
...
+                "net1": ColwiseParallel(),
+                "net2": RowwiseParallel(),
+                "net3": ColwiseParallel(),
+            }
+            for layer in model:
+                parallelize_module(layer, tp_mesh, parallelize_plan)
+            return model
+
+        # Attach to a schedule
+        if issubclass(ScheduleClass, PipelineScheduleSingle):
```
Comment (wconstab) [inline test/distributed/_composable/test_composability/test_pp_composability.py:459]: note: not in this PR, but we might want to clean this pp init stuff up later.  we should get rid of the 'single' vs 'multi' base classes and make it easier to construct a schedule without so many lines of boilerplate.  Or maybe i'm wrong and single vs multi is legitimately worth having different init flows for?? cc @H-Huang

**Review by wconstab — APPROVED** [review]
overall LGTM. thanks for all the improvements! Can you check with @awgu  and @kwen2501 also for any more combinations of pp/fsdp that are important to include in the testing?

One follow up i think is to add fp8.  It can probably be an extension of this test function with one more parameterization, and in another PR.

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 9]
```diff
@@ -357,6 +377,179 @@ def _dcp_test(self):
...
+            mesh_dim_names=("dp", "pp", "tp"),
+        )
+        dp_mesh = device_mesh["dp"]
+        tp_mesh = device_mesh["tp"]
+        pp_mesh = device_mesh["pp"]
+        pp_group = device_mesh["pp"].get_group()
+
+        # create "entire model"
+        total_layers = 8
+        dim = 8
```
Comment (kwen2501) [inline test/distributed/_composable/test_composability/test_pp_composability.py:426]: nit: consider using `dim` in your model definition?

File: test/distributed/_composable/test_composability/test_pp_composability.py [hunk 10]
```diff
@@ -58,6 +67,20 @@ def forward(self, x):
         return x
 
 
+class MLPModuleEven(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.net1 = nn.Linear(8, 8)
+        self.net2 = nn.Linear(8, 8)
+        self.net3 = nn.Linear(8, 16)
```
Comment (kwen2501) [inline test/distributed/_composable/test_composability/test_pp_composability.py:75]: To answer your q, here you have to use 16 because of the colwise you apply to net3.

**Review by kwen2501 — APPROVED** [review]
LGTM.
